urlff74b96b0d_HugoFeng_HospitalProject_tgz-pJ8-hugo_project_hospital_ProjectTestJ8/src/hugo/project/hospital/Hospital.java

```diff
@@ -0,0 +1,655 @@
+		departmentMap.addEdge(getDepartmentByName("Radiotherapy"), getDepartmentByName("Oncology"));
+		departmentMap.addEdge(getDepartmentByName("Oncology"), getDepartmentByName("Cardiology"));
+		departmentMap.addEdge(getDepartmentByName("Oncology"), getDepartmentByName("Radiotherapy"));
+		departmentMap.addEdge(getDepartmentByName("Oncology"), getDepartmentByName("Physiotherapy"));
+		departmentMap.addEdge(getDepartmentByName("Physiotherapy"), getDepartmentByName("Oncology"));
+	}
+	public Department getDepartmentByName(String depName) {
+		for (Department department : departmentList) {
+			if(department.getName().equals(depName)) return department;
+		}
+		return null;
+	}
+	
+	protected Device deviceFactory(String type, int initialAmount) {
+		if(type.equals("needle")) return new NeedleDevice(initialAmount);
+		if(type.equals("stethoscope")) return new StethoscopeDevice(initialAmount);
+		return null;
+	}
+	
+	public String getRoomNumberForPatientNamed(String name) {
+		return reception.getPatientInfo(name).getRoom().toString();
+	}
+	
+	public void signInPatientToDepartment(Patient p, String depName) {
+		Department department = getDepartmentByName(depName);
+		department.addPatient(p);
+	}
+	
+	
+	public void signOutPatient(Patient p) {
+		signOutPatient(p.getName());
+	}
+	
+	public void signOutPatient(String name) {
+		PatientInfo info = reception.getPatientInfo(name);
+		Department department = info.getDepartment();
+		department.signOutPatient(info.getPatient());
+		reception.removePatientInfo(info);
+	}
+	
+	public void printRouteFromTo(String fromDepartment, String toDepartment) {
+		LinkedList<Department> path = departmentMap.findPath(
+				getDepartmentByName(fromDepartment), 
+				getDepartmentByName(toDepartment));
+		if (path == null) {
+			System.out.println("Route from " + fromDepartment + " to " + toDepartment + " not found!");
+			return;
+		} else {
+			String outputString = "Path from " + fromDepartment + " to " + toDepartment + ":\n"
+					+ "\t" + fromDepartment;
+			for (Department department : path) {
+				outputString += " >> " + department.getName();
+			}
+			System.out.println(outputString);
+		}
+	}
+	
+	public void printRouteFromToAvoiding(String fromDepartment, String toDepartment, String avoidDepartment) {
+		LinkedList<Department> path = departmentMap.findPathAvoiding(
+				getDepartmentByName(fromDepartment), 
+				getDepartmentByName(toDepartment),
+				getDepartmentByName(avoidDepartment)); 
+
+		if (path == null) {
+			System.out.println("Route from " + fromDepartment 
+					+ " to " + toDepartment 
+					+ " avoiding " + avoidDepartment
+					+ " not found!");
+			return;
+		} else {
+			String outputString = "Path from " + fromDepartment
+					+ " to " + toDepartment 
+					+ " avoiding " + avoidDepartment + ":\n"
+					+ "\t" + fromDepartment;
+			for (Department department : path) {
+				outputString += " >> " + department.getName();
+			}
+			System.out.println(outputString);
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 590:605, 614:629.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.